src/backend/distributed/commands/function.c

```diff
@@ -415,15 +396,31 @@ EnsureFunctionCanBeColocatedWithTable(Oid functionOid, Oid distributionColumnTyp
 								"\"citus.replication_model\" is set to \"streaming\"")));
 	}
 
+	/*
+	 * If the types are the same, we're good. If not, we still check if there
+	 * is any coercion path between the types.
+	 */
 	sourceDistributionColumnType = sourceDistributionColumn->vartype;
 	if (sourceDistributionColumnType != distributionColumnType)
 	{
-		char *functionName = get_func_name(functionOid);
-		char *sourceRelationName = get_rel_name(sourceRelationId);
+		Oid coercionFuncId = InvalidOid;
+		CoercionPathType coercionType = COERCION_PATH_NONE;
 
-		ereport(ERROR, (errmsg("cannot colocate function \"%s\" and table \"%s\" "
-							   "because distribution column types don't match",
-							   sourceRelationName, functionName)));
+		coercionType =
+			find_coercion_pathway(sourceDistributionColumnType, distributionColumnType,
```

Comment:
Yes, casts are one way functions.

Reply:
An example I could find:

```
postgres=# select '1'::jsonb::int;
 int4 
------
    1
(1 row)

postgres=# select '1'::int::jsonb;
ERROR:  cannot cast type integer to jsonb
LINE 1: select '1'::int::jsonb;

```

Reply:
So, then I should change the order such that we are compatible with [here](https://github.com/citusdata/citus/pull/2965/files#diff-82b34f694db38c26921dbd5d12890974R131) ?

Reply:
ah, I misses that `distributionColumnType` is the distribution argument type rather than the distribution column type.

Yes, we should go distribution argument type -> distribution column type.